compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUAttrs.td

```diff
@@ -0,0 +1,120 @@
+
+include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUDialect.td"
+include "iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUInterfaces.td"
+include "mlir/IR/OpBase.td"
+include "mlir/IR/AttrTypeBase.td"
+include "mlir/IR/EnumAttr.td"
+
+def IREEGPU_MmaArrayAttr : TypedArrayAttrBase<IREEGPU_MmaAttr,
+    "Descriptor array of a list of supported mma operations">;
+
+class IREEGPU_MmaVectorLayoutAttr<string attrname, string mmatype> :
+    AttrDef<IREEGPU_Dialect, attrname, [
+  DeclareAttrInterfaceMethods<IREEGPU_MmaAttr,[
+    "getABCElementTypes",
+    "getContractionLayout",
+    "getMNKShape",
+  ]>
+]> {
+  let cppNamespace = "::mlir::iree_compiler::IREE::GPU";
+
+  string baseDescription = [{
+    Attribute describing a particular shape of matrix-multiply and accumulate
+    instruction. Abstractly, all operations of this type represent the following
+    unit of arithmetic for matrices A, B, and C.
+
+    ```
+      C += A x B
+    ```
+
+    Where the shape of matrix `A` is `[m, k]`, `B` is `[k, n]`, and
+    `C` is `[m, n]`. This intentionally leaves the layout information abstract
+    and uses interface methods to materialize layout information only when
+    needed. The shape of the mma is stored explicitly here as that information
+    is queried frequently.
+
+    The element types for this particular mma type are |aType|, |bType|, and
+    |cType| for matrices `A`, `B`, and `C` respectively.
+
+    ######
+
+  }];
+
+    
+  let parameters = (ins
+    mmatype:$type,
+    "int64_t":$mSize,
+    "int64_t":$nSize,
+    "int64_t":$kSize,
+    "::mlir::Type":$aType,
+    "::mlir::Type":$bType,
+    "::mlir::Type":$cType
+  );
+
+  let extraClassDeclaration = [{
+    ::std::tuple<Type, Type, Type> getABCElementTypes() {
+      return ::std::make_tuple(getAType(), getBType(), getCType());
+    }
+
+    ::std::tuple<int64_t, int64_t, int64_t> getMNKShape() {
+      return ::std::make_tuple(getMSize(), getNSize(), getKSize());
+    }
+  }];
+}
+
+class IREEGPU_I32MmaEnumAttr<string name, string summary, list<I32EnumAttrCase> cases>
+    : I32EnumAttr<name, summary, cases> {
+  let cppNamespace = "::mlir::iree_compiler::IREE::GPU";
+  let genSpecializedAttr = 0;
+}
+
+class IREEGPU_MmaEnumAttr<EnumAttrInfo enumInfo, string name = "">
+  : EnumAttr<IREEGPU_Dialect, enumInfo, name>;
+
+//===----------------------------------------------------------------------===//
+// MFMA Attributes
+//===----------------------------------------------------------------------===//
+
+def F16_16x16x16_F32 : I32EnumAttrCase<"F16_16x16x16_F32", 0>;
+def F16_32x32x8_F32 : I32EnumAttrCase<"F16_32x32x8_F32", 1>;
+
+def IREEGPU_MFMAType : IREEGPU_I32MmaEnumAttr<"MFMAType",
+    "Descriptor for different MFMA operations", [
+      F16_16x16x16_F32,
+      F16_32x32x8_F32,
+    ]>; 
+
+def IREEGPU_MFMATypeAttr : IREEGPU_MmaEnumAttr<IREEGPU_MFMAType, "mfma_type">;
+
+def IREEGPU_MFMA : IREEGPU_MmaVectorLayoutAttr<"MFMA", "MFMATypeAttr"> {
```

Comment:
Thanks for the review, will get to the other comments but wanted to respond to this first. My thinking was that we'd have a different attribute type for `mfma`, `wmma_amd`, `wmma_nv`, and `mma.sync` because the lowering targets for those attributes are different, so it's nice to do the pattern population for them using `dyn_cast<MFMAAttr>` or `dyn_cast<WMMAAttr>`. Then each of those would have different variants.

So in short, definitely agree we don't want too much code duplication for all of the layout variants. Adding the target chip to `getContractionLayout` sounds good in case different architectures have different explicit layouts so I'll do that, but I think keeping the split between the top level intrinsic types of mfma/wmma/mma_sync/etc. simplifies the later lowering logic a bit.

Reply:
Yeah that might be a good balance for now. hiding the chip aspect behind `getContractionLayout` is already an improvement. we can iterate on it as we go.